### ISSUE TITLE
Failed to compile fragment shader when batch size one greater than WEBGL_MAX_TEXTURES_IN_SHADER

### ISSUE THREAD

**Issue by gnodar01** [tensorflow/tfjs#6884]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow.js): **Yes**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): multiple versions of macOS, 10.15 through 12.6, on machines with both intel and m1 chips (have not tried on windows). I've reproduced the error in a codepen (linked below), which I guess runs in a linux container. This problem is webgl specific though, I haven't tried server-side in a node backend
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: **Yes** (iphone X)
- TensorFlow.js installed from (npm or script link): **both** (originally encountered from npm installed version, but was able to reproduce in script install (linked below)
- TensorFlow.js version (use command below): **3.20.0**
- Browser version: **Chrome, latest; Firefox, latest**
- Tensorflow.js Converter Version: **n/a**

**Describe the current behavior**

This problem occurs when using `fitDataset`, but as far as I can tell does not happen on `fit`.

Using the webgl backend, any time batch size is exactly 1 greater than a multiple of the returned value of `env().getNumber('WEBGL_MAX_TEXTURES_IN_SHADER')` (in my case 16), I receive a "Failed to compile fragment shader" error, along with 200 lines of GLSL source code with the error "ERROR: 0:176: 'getT1' : no matching overloaded function found" .

For instance setting batch size directly to 33, or setting it to 32 but only having 17 samples in the final batch of the training or validation sets.

**Describe the expected behavior**

Specific numbers where `currentBatchSize % 16 == 1`, shouldn't cause errors.

**Standalone code to reproduce the issue**
I've reproduced the error as simply as I could, by modifying the mnist example project to use datasets, along with `fitDataset`, here: https://codepen.io/gnodar01/pen/yLjjgVQ

I've set batch size to 497, which produces the error since `497 % 16 == 1`, but setting to 496 or 498 produces no error.

**Other info / logs**

I originally encountered this behavior in a web application we're building, with more complex code. I walk through and describe where I think the problem occurs in tfjs here: https://github.com/piximi/piximi/issues/385#issuecomment-1264020896.

The short version is that `concatImpl` recrusively calls itself with subarrays of size 16 or less. When one of those recursive calls contains an input array of size one, the error occurs.

The codepen I've linked to reproduces the error when you click the "RUN" button and shows the stacktrace

**Comment by Linchenn** [comment]
Thank you for locating this error! I think it is because WebGL backend's tf.concat could not handle the input array of size 1. Will send a fix soon.

**Comment by Linchenn** [comment]
Just merged the fix. When using new build of [tf-backend-webgl.js](https://github.com/Linchenn/Toolkit/blob/main/debugTFJS/issue6884/tf-backend-webgl-fix.js), this error disappeared.

**Comment by gnodar01** [comment]
Thanks @Linchenn
